Map Icons with IDs

jagannatharjun opened the pull request:
- [X] Map icons in `qbt-theme` folder
- [x] Map icons in `skins` folder
- [x]  Make System Icon theme work

glassez commented on src/icons/icons.qrc:
```diff
@@ -249,6 +249,7 @@
         <file>flags/za.svg</file>
         <file>flags/zm.svg</file>
         <file>flags/zw.svg</file>
+        <file>icons.json</file>
```
I would like more specific name, e.g. "iconmapping.json" or "iconconfig.json"

LordNyriox replied:
@jagannatharjun:  IMO, `iconmapping.json` sounds more descriptive.

glassez replied:
Or just `config.json` if it is supposed to be in `icons` directory.

LordNyriox commented on src/icons/icons.json:
```diff
@@ -0,0 +1,159 @@
+    "PeerList.ContextMenu.AddAction": "qbt-theme/user-group-new",
+    "PeerList.ContextMenu.BanAction": "qbt-theme/user-group-delete",
+    "PeerList.ContextMenu.CopyAction": "qbt-theme/edit-copy",
+    "Rss": "qbt-theme/inode-directory",
```
@jagannatharjun: 

>RSS 

As in, use all caps in this one case.

glassez replied:
I mean all `RSS` occurrences in icon IDs.

LordNyriox commented:
@jagannatharjun:  After this PR is finalized and merged, would you be willing to post a template theme (including icons) for qBittorrent?

jagannatharjun commented:
@LordNyriox yes

glassez commented on src/gui/aboutdialog.cpp:
```diff
@@ -45,7 +45,7 @@ AboutDialog::AboutDialog(QWidget *parent)
     m_ui->labelName->setText(QString("<b><h2>qBittorrent " QBT_VERSION " (%1-bit)</h2></b>").arg(QT_POINTER_SIZE * 8));
 
-    m_ui->logo->setPixmap(Utils::Gui::scaledPixmapSvg(":/icons/skin/qbittorrent-tray.svg", this, 32));
+    m_ui->logo->setPixmap(Utils::Gui::scaledPixmapSvg("QBittorrent.Logo", this, 32));
```
I would use "AboutDialog.Logo".

LordNyriox replied:
@glassez:  It is my understanding that the same icon is used in multiple places (including the system tray). 

Do you really want to name the icon based on only one of those components?

glassez replied:
@LordNyriox, if you had followed #10903, you might have understood my intentions...
https://github.com/qbittorrent/qBittorrent/issues/10903#issuecomment-510820381

jagannatharjun commented:
Should I fall back to the default icon mapping if the icon path can't be resolved with custom `iconconfig.json` . May be provide an option

I am thinking to replace the current selection method with something like [this](https://github.com/qbittorrent/qBittorrent/pull/10702#issuecomment-496570342).
With the current setup, It seems more natural.

glassez commented:
My main suggestions for naming are the following:
ID should contain two main parts.
First part (actually last part of ID) is "common semantic" (e.g. "AddAction" or "Logo").
The second part is "scope" of particular ID, possible multilevel (e.g. "AddTorrentDialog.TorrentContentView").
"Scope" can be omitted in icon mapping (partially or fully) by replacing with "*" character. It allows the theme creator to achieve any desired detail. E.g. someone can set different icons for each particular ID but someone else can create very reduced theme (by having only entries like "*.AddAction", i.e. set the same icon for all similar places).
We should avoid any "internals" in ID parts. I.e. "ContentTree" is not so good, IMO (due to "Tree" in it). It should just point to some widget (or view) that shows content of torrent (regardless of the particular type of this widget), so "TorrentContentView" is better name.
We can avoid too much detail in ID scopes, but only if it is not likely to be needed in the future (because we will not be able to change existing IDs without breaking existing themes).

glassez commented:
>Should I fall back to the default icon mapping if the icon path can't be resolved with custom `iconconfig.json`

What do you mean exactly?

If there is no such ID (or appropriate generic ID that matches it) in theme then you should use built-in icon (and cache it for this ID). This situation can occur when some new qBittorrent version introduces new ID so there is no matches in existing themes. You also should indicate this event in some way (maybe show message box about missing icons or some less annoying way). Or maybe don't fallback and just show some predefined icon placeholder?

If the theme contains some ID but mapped path is invalid then it's real bug in theme and you shouldn't fallback to correct icon.

glassez commented on src/gui/uithememanager.cpp:
```diff
@@ -140,5 +156,41 @@ QString UIThemeManager::getIconPath(const QString &iconId) const
+        iconIdPattern = "*" + (sepIndex == -1 ? "" : iconIdPattern.right(iconIdPattern.size() - sepIndex));
+        const auto iter = m_iconMap.find(iconIdPattern);
+        if (iter != m_iconMap.end())
+            iconPath = iter->toString();
```
You should check for empty icon path here since it is error case.
Also you should cache icon path for given icon id in m_iconMap to avoid searching by pattern next time.

glassez commented on src/gui/uithememanager.cpp:
```diff
@@ -140,5 +156,41 @@ QString UIThemeManager::getIconPath(const QString &iconId) const
 #endif
-    return IconProvider::instance()->getIconPath(iconId);
+
+    const auto iter = m_iconMap.find(iconId);
```
Seems this method has logical errors. I need to get a fresh look at it.

jagannatharjun commented on src/gui/uithememanager.cpp:
```diff
@@ -58,13 +60,26 @@ void UIThemeManager::initInstance()
-    if (pref->useCustomUITheme()
+    const bool useCustomUITheme = pref->useCustomUITheme();
+    if (useCustomUITheme
         && !QResource::registerResource(pref->customUIThemePath(), "/uitheme"))
```
only here? what if the resource file doesn't have `iconconfig.json` or when some `iconid` can't be found in the custom `iconconfig.json`

Chocobo1 commented on src/gui/uithememanager.cpp:
```diff
@@ -109,19 +130,20 @@ QIcon UIThemeManager::getIcon(const QString &iconId, const QString &fallback) co
+    return QIcon(m_iconsDir + "flags/" + countryIsoCode.toLower() + ".svg");
 }
 
 QString UIThemeManager::getIconPath(const QString &iconId) const
```
Do the code inside `if (m_useSystemTheme) {}` still work? It is important for linux users to use native icons.

glassez replied:
>It is important for linux users to use native icons.

As I understand it, "native icons" should be used (if enabled) no matter what theme is selected, built-in or custom?

Chocobo1 replied:
>As I understand it, "native icons" should be used (if enabled) no matter what theme is selected, built-in or custom?

I think yes.

jagannatharjun replied:
Dor this we can have another config file with system icons as mapped value

glassez replied:
I think something like that will be needed...

jagannatharjun replied:
ok, I tried but didn't test. Actually, I don't have any Linux computer available and never used one either so it will take me some time to set up one. Can anybody else test this?

jagannatharjun commented:
![image](https://user-images.githubusercontent.com/34717789/63773807-942e8780-c8f9-11e9-9323-3f20e7c1bd51.png)
@glassez  update?

glassez commented:
>@glassez update?

I'm sorry, I just can't get around to it for lack of free time. It's on my to-do list, so I'll be back here as soon as I can.

glassez commented on src/gui/uithememanager.cpp:
```diff
@@ -58,13 +60,35 @@ void UIThemeManager::initInstance()
+
+    m_iconsDir = useCustomUITheme ? ":uitheme/icons/" : ":icons/";
+
+    QFile iconJsonMap(m_iconsDir + iconConfigFile);
```
I'm not sure that's right... Does it make sense to override mapping for "system theme" icons in a custom theme? Shouldn't there just be a three-state option (default-system-custom)? @Chocobo1?

Chocobo1 replied:
>Does it make sense to override mapping for "system theme" icons in a custom theme?

I think not.

>Shouldn't there just be a three-state option (default-system-custom)?

I agree although I haven't checked how the GUI should look like.

glassez commented on src/gui/uithememanager.cpp:
```diff
@@ -99,29 +123,30 @@ QIcon UIThemeManager::getIcon(const QString &iconId, const QString &fallback) co
     if (m_useSystemTheme) {
-        QIcon icon = QIcon::fromTheme(iconId);
+        QIcon icon = QIcon::fromTheme(m_iconMap[iconId].toString());
         if (icon.name() != iconId)
```
You should compare with mapped icon id to follow current logic. Otherwise it's meaningless.

glassez commented on src/gui/aboutdialog.cpp:
```diff
@@ -45,7 +45,7 @@ AboutDialog::AboutDialog(QWidget *parent)
     m_ui->labelName->setText(QString("<b><h2>qBittorrent " QBT_VERSION " (%1-bit)</h2></b>").arg(QT_POINTER_SIZE * 8));
 
-    m_ui->logo->setPixmap(Utils::Gui::scaledPixmapSvg(":/icons/skin/qbittorrent-tray.svg", this, 32));
+    m_ui->logo->setPixmap(Utils::Gui::scaledPixmapSvg("AboutDialog.Logo", this, 32));
```
I strongly dislike it in this way.
You need to merge `Utils::Gui::scaledPixmapSvg()` and `Utils::Gui::scaledPixmap()`. Also the first parameter shouldn't be changed to iconId. You just should to call it as `Utils::Gui::scaledPixmapSvg(UIThemeManager::getIconPath("AboutDialog.Logo"), this, 32)`

glassez replied:
Or (even better, IMO) keep `getIconPath()` private and implement `UIThemeManager::getPixmap()` that encapsulates all needed logic.

jagannatharjun replied:
Did you mean by following?
```
QPixmap scaledPixmap(const QPixmap& pixmap, const QWidget *widget, int height = 0);
QPixmap scaledPixmapSvg(const QPixmap &pixmap, const QWidget *widget, int baseHeight);
```

If yes, then `scaledPixmapSvg` requires to have some key to cache, so will it be better to move b
`scaledPixmap` and `scaledPixmapSvg` both inside the `UIThemeManager`.

jagannatharjun replied:
Did you mean by following?
```
QPixmap scaledPixmap(const QPixmap& pixmap, const QWidget *widget, int height = 0);
QPixmap scaledPixmapSvg(const QPixmap &pixmap, const QWidget *widget, int baseHeight);
```

If yes, then `scaledPixmapSvg` requires to have some key to cache, so will it be better to move b
`scaledPixmap` and `scaledPixmapSvg` both inside the `UIThemeManager`.

glassez replied:
I mean you should implement only one of the following:
```c++
QPixmap UIThemeManager::getPixmap(const QString &id, const QWidget *widget, int baseHeight);
QPixmap Utils::Gui::scaledPixmap(const QPixmap &pixmap, const QWidget *widget, int baseHeight);
```

The client code should know nothing about the type of some particular icon/pixmap (it can be different in different theme). So the function should determine it itself.

jagannatharjun replied:
I tried implementing `UIThemeManager::getPixmap` because then everything will be in one place.I implement this on the basis of scaledPixmapSvg. Can you check now?

glassez commented on src/gui/uithememanager.cpp:
```diff
@@ -99,29 +123,30 @@ QIcon UIThemeManager::getIcon(const QString &iconId, const QString &fallback) co
+        QIcon icon = QIcon::fromTheme(m_iconMap[iconId].toString());
         if (icon.name() != iconId)
-            icon = QIcon::fromTheme(fallback, QIcon(IconProvider::instance()->getIconPath(iconId)));
+            icon = QIcon::fromTheme(fallback, QIcon(getIconPath(iconId)));
```
Also how do you suppose to provide fallback icon now? It seems that it should come from (system theme) icon mapping now.

jagannatharjun replied:
Should I add separate icon id for fallback as well? I think there are only like 4 instances where fallback is actually used.

Well, It does make sense to add IconIds for these as well. Will replace direct references with Icon ids.

glassez commented on src/gui/uithememanager.cpp:
```diff
@@ -99,29 +123,30 @@ QIcon UIThemeManager::getIcon(const QString &iconId, const QString &fallback) co
 {
     if (countryIsoCode.isEmpty()) return {};
-    return QIcon(":/icons/flags/" + countryIsoCode.toLower() + ".svg");
+    return QIcon(m_iconsDir + "flags/" + countryIsoCode.toLower() + ".svg");
```
Shouldn't we cache flag icons too?

jagannatharjun replied:
Done.

glassez commented on src/gui/uithememanager.cpp:
```diff
@@ -140,5 +165,43 @@ QString UIThemeManager::getIconPath(const QString &iconId) const
+    }
+
+    if (iconPath.isEmpty()) {
+        LogMsg(tr("Can't resolve icon id - %1").arg(iconId), Log::WARNING);
```
Since missing ID isn't really error case (as I said before this case can occur when new ID is introduced in some qBittorrent version but theme is outdated yet) you should fallback to default theme icon.

The real error cases are:
1. Empty filename
2. File doesn't exist

IMO, we should reject such invalid theme at all (and show message to the user).

I also think you can check theme for these errors once it is loaded (since loaded theme is immutable).

jagannatharjun replied:
Do you mean by  - for every `iconId` which doesn't exist in iconMap, try finding that in defaultIconMap which means - 
1. Load both default iconconfig and custom iconconfig (if custom is enabled) in the constructor
2. Reject custom theme if "real error case" exists, (should I check the default Theme for errors too, but only show the error)
3. In `getIconPath`, for every `iconId` which doesn't exist in iconMap, try finding that in defaultIconMap

glassez replied:
>for every iconId which doesn't exist in iconMap, try finding that in defaultIconMap

Generally, Yes.
But I think there is a better algorithm:
1. Load default iconconfig. Check it and show errors (or maybe don't check it at all?).
2. Load custom iconconfig. Check it and reject if "real error case" exists
3. Join both iconconfigs (add unique and replace existing values in default config with the ones from custom config.
4. Use resulting icon config.

jagannatharjun replied:
While checking should I load the icon directly i.e store the icons in something like `std::map<QString, QIcon>`,

Chocobo1 commented on src/gui/uithememanager.h:
```diff
@@ -48,10 +48,11 @@ class UIThemeManager : public QObject
     QIcon getIcon(const QString &iconId, const QString &fallback) const;
     QIcon getFlagIcon(const QString &countryIsoCode) const;
-    QString getIconPath(const QString &iconId) const;
+    QPixmap getPixmap(const QString &iconId, const QWidget *widget, int baseHeight) const;
```
Does it make sense that a `getPixmap` operation having a `QWidget` as one of the parameters? or maybe a better function name is in need?

glassez replied:
>or maybe a more precise function name is in need?

getScaledPixmap?

Chocobo1 commented on src/gui/utils.cpp:
```diff
@@ -76,35 +75,6 @@ qreal Utils::Gui::screenScalingFactor(const QWidget *widget)
 #endif // Q_OS_WIN
 }
 
-QPixmap Utils::Gui::scaledPixmap(const QIcon &icon, const QWidget *widget, const int height)
```
I haven't been following closely to this PR, so pardon if it doesn't make sense in advance.
I think merging these functions within existing uithememanager methods is a bad idea, those are image scaling functions, it should be preserved in it own form (private functions in uithememanager is acceptable for me, and making it public when required). See how you load `mascot.png`, spilling all the unnecessary details of image scaling details into it.

glassez replied:
>I haven't been following closely to this PR

UIThemeManager should provide icon (or pixmap) by its ID.
The client code (of UIThemeManager) should know nothing about theme details (e.g. real file types etc.).
In some places we need to get scaled pixmap. Currently we have two helpers to scale pixmap (depending on its file type). But we can't proceed to use this way considering the above. So I suggested moving this logic inside the UIThemeManager class. Otherwise we can leave only one helper that will scale existing pixmap (not so optimal since we can't use SVG optimization).

Chocobo1 replied:
I think this can work (a careless sketch up):
```c++
QPixmap UIThemeManager::getScaledPixmap(const QString &iconId, const QWidget *widget, int baseHeight) const
{
    const QString iconPath = UIThemeManager::instance()->getIconPath(iconId);
    return Utils::Gui::scaledPixmap(iconPath, widget, baseHeight);  // IMO this function has nothing to do with theme manager
}

QPixmap Utils::Gui::scaledPixmap(const QString &iconPath, const QWidget *widget, const int baseHeight)
{
    const int scaledHeight = baseHeight * Utils::Gui::screenScalingFactor(widget);
    const QString normalizedKey = iconPath + '@' + QString::number(scaledHeight);

    QPixmap pm;
    QPixmapCache cache;
    if (!cache.find(normalizedKey, &pm)) {
        pm = QIcon(iconPath).pixmap(scaledHeight);     
        cache.insert(normalizedKey, pm);
    }
    return pm;
}
```

glassez replied:
>I think this can work (a careless sketch up):

I think, Yes.
But UIThemeManager instance shouldn't call itself indirectly:
```c++
QPixmap UIThemeManager::getScaledPixmap(const QString &iconId, const QWidget *widget, int baseHeight) const
{
    const QString iconPath = getIconPath(iconId);
    return Utils::Gui::scaledPixmap(iconPath, widget, baseHeight);
}
```

Also I wouldn't have `widget` as param but `scalingFactor` to be more abstract.

jagannatharjun replied:
now?

jagannatharjun replied:
Now I have reset the changes inside `utils.cpp`.  I don't do anything with `utils.*`.  Changed `getPixmap` to `getScaledPixmap` which calls `Utils::Gui::scaledPixmap` as you guys said.

Chocobo1 commented on src/gui/utils.cpp:
```diff
@@ -76,35 +75,6 @@ qreal Utils::Gui::screenScalingFactor(const QWidget *widget)
 #endif // Q_OS_WIN
 }
 
-QPixmap Utils::Gui::scaledPixmap(const QIcon &icon, const QWidget *widget, const int height)
```
There is also a second consideration to be discussed, is it correct/preferable that a theme manager needs to be concerned about image scaling? Sure it simplifies the code using it, but the big picture looks weird to me. Of course we can have helper functions that provide such conveniences but they should be static functions of the theme manager class.

jagannatharjun commented on src/gui/uithememanager.cpp:
```diff
@@ -58,13 +60,35 @@ void UIThemeManager::initInstance()
+        LogMsg(tr("Failed to load UI theme from file: \"%1\", falling back to system default theme").arg(pref->customUIThemePath()), Log::WARNING);
+
+        useCustomUITheme = false;
+        Preferences::instance()->setUseCustomUITheme(false);
```
`UIThemeManager::applyStylesheet` will fail then.

jagannatharjun replied:
I mean `UIThemeManager::applyStylesheet` should be aware of the error, I can have an internal variable.

glassez replied:
>I can have an internal variable.

:+1:

jagannatharjun replied:
Now there is an internal variable `m_useCustomStylesheet` , loads custom stylesheet if available, not resets the preferences anymore.

glassez commented on src/gui/uithememanager.h:
```diff
@@ -48,13 +48,20 @@ class UIThemeManager : public QObject
     QIcon getIcon(const QString &iconId, const QString &fallback) const;
     QIcon getFlagIcon(const QString &countryIsoCode) const;
+    QPixmap getScaledPixmap(const QString &iconId, const QWidget *widget, int baseHeight) const;
+    QString getIconPath(const QString &iconId) const;
```
Ideally it should be private. Is there any reason to keep it public?

jagannatharjun replied:
It's actually used to provide icon path to html

https://github.com/jagannatharjun/qBittorrent/blob/d799c7df2029cd17c134d2e66cda874f78bf8af9/src/gui/mainwindow.cpp#L1524-L1536

jagannatharjun replied:
https://github.com/jagannatharjun/qBittorrent/blob/d799c7df2029cd17c134d2e66cda874f78bf8af9/src/gui/mainwindow.cpp#L1524-L1536

I think the above places require hardcoded path anyway, so I will remove the changes and make `getIconPath` private.

glassez replied:
>I think the above places require hardcoded path anyway

Why?

jagannatharjun replied:
okay sorry!

glassez commented on src/gui/uithememanager.h:
```diff
@@ -48,13 +48,20 @@ class UIThemeManager : public QObject
-    QString getIconPath(const QString &iconId) const;
+    using IconMap = QHash<QString, QString>;
+
+    static bool loadIconConfig(const QString &configFile, const QString &iconDir, IconMap &config);
```
I would make it non-static and drop last parameter.

glassez replied:
On the other hand, it should be really class-independent (and it is actually used to load different files). Then just put it in an anonymous namespace in .cpp. You will also have to change the error handling in it (throw Runtime Error in it and print log message in calling code).

IMO, using static methods in C++ is unjustified in most cases.

glassez commented on src/gui/uithememanager.cpp:
```diff
@@ -32,14 +32,84 @@
+#include "utils.h"
 
 UIThemeManager *UIThemeManager::m_instance = nullptr;
+static const char *allowedExts[] = {".svg", ".png"};
```
I would create QString constants to omit repeated conversions.

jagannatharjun commented:
with sudo
![image](https://user-images.githubusercontent.com/34717789/65025720-d7a66f80-d954-11e9-919d-fad4821eda44.png)

without sudo
![image](https://user-images.githubusercontent.com/34717789/65025639-b8a7dd80-d954-11e9-864f-8475b5da898d.png)

🤔

jagannatharjun commented:
I think with the addition of `search-dirs`, the qbittorrent's `iconconfig.json` and `systemiconconfig.json` can be merged into a single file.

What do you think @glassez  @Chocobo1  ?

glassez commented:
>I think with the addition of `search-dirs`, the qbittorrent's `iconconfig.json` and `systemiconconfig.json` can be merged into a single file.
What do you think @glassez  @Chocobo1?

Sorry, I can't keep up with you...
What's `search-dirs` for?

jagannatharjun commented:
> What's `search-dirs` for?

I introduced this in the previous commit to making system icon theme works because there may be some icons that are not present in the system theme. So I added an extra parameter in JSON `search-dirs` which specifies the directory inside the uitheme to look for the icon.

This can be used to merge the systemiconconfig.json and iconconfig.json as most of the system icons have the same name. Merging means providing an alias in qrc, this way user can still specify a custom `systemiconconfig.json`

jagannatharjun commented:
actually with current PR state, we can actually use `systemiconconfig.json` on all systems as default icon config and system icon theme will only be used when allowed by preferences.

glassez commented:
>I introduced this in the previous commit to making system icon theme works because there may be some icons that are not present in the system theme.

It seems that you went beyond the designed model. Besides, it required you to introduce new entities (i.e. "search dirs").
When some icon isn't presented in current system theme you should just use default icon for this IconID (that are supposed to always be presented).

jagannatharjun commented:
> When some icon isn't presented in current system theme you should just use default icon for this IconID (that are supposed to always be presented).

So in systemiconconfig.json, all values should be treated as system icons? don't allow paths?

Also, I don't think `m_allowedExt` is actually needed, I will be removing it and replacing the icons with full qualified icon paths in iconconfig.jsons

glassez commented:
>Also, I don't think `m_allowedExt` is actually needed, I will be removing it and replacing the icons with full qualified icon paths in iconconfig.json

:+1:
>So in systemiconconfig.json, all values should be treated as system icons?

Yes, of course.

Also I would still move system and default mapping into the code.

jagannatharjun commented:
> Yes, of course.

Well, will following be ok? -
1. Load both systemiconconfig.json and iconconfig.json. There should be no check for systemiconconfig.json
2. in geticonpath, if the QIcon::fromTheme returned an empty icon then get the icon from the `iconconfig.json`

>   Also I would still move system and default mapping into the code.

separately in `iconconfig.h` or inside the `uithememanager.h`

jagannatharjun commented:
Oh! 😫  I was implementing this and It doesn't make sense XD.
my final take will be:
1. Load custom systemiconconfig and merge with default iconconfig
2. Load custom iconconfig and merge with default iconconfig
3. In getIcon if QIcon::fromThem and Fallback both fails, get the icon from iconconfig

jagannatharjun commented on src/gui/uithememanager.cpp:
```diff
@@ -55,16 +150,49 @@ void UIThemeManager::initInstance()
+            m_flagsDir = ":uitheme/icons/flags/";
+
+            // by chaining following functions, we ignore the erroed values in customIconMap
+            customIconMap = resolveDirs(loadIconConfig(":uitheme/icons/iconconfig.json"), ":uitheme/icons/");
```
Maybe ask the custom theme for absolute icon path, this way they are not just limited to qrc but reference icon from somewhere else also.

jagannatharjun commented:
![image](https://user-images.githubusercontent.com/34717789/65382179-c4145380-dd1c-11e9-8688-ce9bce612d4c.png)

ubuntu icon theme - [link](https://github.com/jagannatharjun/qbt-theme/blob/master/ubuntu/style.qbtheme?raw=true)

glassez commented on src/gui/uithememanager.cpp:
```diff
@@ -55,16 +138,56 @@ void UIThemeManager::initInstance()
+                throw ThemeError(tr("Failed to load UI theme from file: \"%1\".")
+                                 .arg(pref->customUIThemePath()));
+            m_useCustomStylesheet = QFile::exists(":uitheme/stylesheet.qss");
+            m_flagsDir = ":uitheme/icons/flags/";
```
Is there any real reason to allow custom flags? Can they be different from the default ones?

jagannatharjun replied:
well I have seen people with custom flags but this will be very cumbersome for any one to do. So let me remove them for now.

jagannatharjun replied:
removed custom `m_flagsdir`

glassez commented on src/gui/iconconfig.cpp:
```diff
@@ -0,0 +1,390 @@
+#include <QHash>
+#include <QString>
+
+extern const QHash<QString, QString> iconConfig = {
```
Since these consts are supposed to be used in one .cpp file I would put them into .h instead and omit extern declaration (or even put them into corresponding .cpp directly).

jagannatharjun replied:
moved in uithememanager.cpp

glassez commented on src/gui/rss/articlelistwidget.cpp:
```diff
@@ -98,7 +99,7 @@ void ArticleListWidget::handleArticleRead(RSS::Article *rssArticle)
 
     item->setData(Qt::ForegroundRole, QPalette().color(QPalette::Inactive, QPalette::WindowText));
-    item->setData(Qt::DecorationRole, QIcon(":/icons/sphere.png"));
+    item->setData(Qt::DecorationRole, UIThemeManager::instance()->getIcon("RSS.ReadArticle"));
```
I would use something like "RSSView" instead of "RSS".

jagannatharjun replied:
May be `RSSTab`?

glassez replied:
Tabs is just a way to organise Views/Widgets in container (Main window in this case).

jagannatharjun replied:
They're actually used in RSS Feed List so may be `RSSFeedList`, but `RSSView` seems good to

jagannatharjun commented on src/gui/uithememanager.cpp:
```diff
@@ -30,17 +30,489 @@
+        // copy jsonObject to config, also check for errors
+        for (auto i = jsonObject.begin(), e = jsonObject.end(); i != e; ++i) {
+            if (!i.value().isString())
+                throw ThemeError(QObject::tr("Error in iconconfig \"%1\", error: Provided value for %2, is not a string.")
```
There are two `iconconfig.json` `systemiconconfig.json`

jagannatharjun replied:
I think message lost its context without `Error in iconconfig \"%1\"`,
maybe replace it with `custom icon config json format error` or maybe something better?

glassez replied:
Maybe `Icon configuration file format error: %1.`?

jagannatharjun commented on src/gui/uithememanager.cpp:
```diff
@@ -55,16 +527,50 @@ void UIThemeManager::initInstance()
+
+#ifdef QBT_SYSTEMICONS
+            if (m_useSystemTheme)
+                m_systemIconMap = loadIconConfig(":uitheme/icons/systemiconconfig.json");
```
no It's loading the custom `systemiconconfig.json`, Unlike `iconconfig.json` I don't merge the default and custom `systemiconconfig.json`

Here it loads the `systeiconconfig.json`

glassez replied:
Didn't we decide to allow no custom "system icon mapping"?
What's the point to reimplement built-in mapping?

glassez replied:
It seems pointless to me to have several different "system theme" mappings. System theme has no different icons for the same thing.

jagannatharjun commented on src/gui/uithememanager.cpp:
```diff
@@ -30,17 +30,489 @@
+        {"MainMenu.PauseAllAction", ":icons/qbt-theme/media-playback-pause.svg"},
+        {"MainMenu.Queue.BottomAction", ":icons/qbt-theme/go-bottom.svg"},
+        {"MainMenu.Queue.DownAction", ":icons/qbt-theme/go-down.svg"},
+        {"MainMenu.RSSTab", ":icons/qbt-theme/application-rss+xml.svg"},
```
Adding `Show` feels like an action, but the icon is used as the Tab Icon.

glassez replied:
>the icon is used as the Tab Icon.

`MainMenu.RSSTab` is used to provide icon for Tab? Then it is incorrect ID.

jagannatharjun replied:
> MainMenu.RSSTab is used to provide icon for Tab?

yeap.

Actually I used this scheme everywhere. f.e `MainMenu.SearchTab`

glassez replied:
MainMenu context is for Main menu entries. For tabs you can use MainWindow context.

jagannatharjun commented on src/gui/aboutdialog.cpp:
```diff
@@ -32,6 +32,7 @@
 
 #include "base/unicodestrings.h"
 #include "base/utils/misc.h"
+#include "gui/uithememanager.h"
```
There are actually only two such reference one is this one you mention and other is `src\gui\rss\articlelistwidget.cpp`. should I don't use this prefix even in `gui\rss` folder?

glassez replied:
Just follow other include names style in the same file.

jagannatharjun commented on src/gui/uithememanager.cpp:
```diff
@@ -30,17 +30,489 @@
+{
+    static const QHash<QString, QString> iconConfig = {
+        {"AboutDialog.Logo", ":icons/skin/qbittorrent-tray.svg"},
+        {"AddNewTorrentDialog.TorrentContentView.RenameAction", ":icons/qbt-theme/edit-rename.svg"},
```
Hmm yes It does seem superfluous.

glassez replied:
I think we should recheck all IDs at last step.

jagannatharjun commented on src/gui/uithememanager.cpp:
```diff
@@ -30,17 +30,489 @@
+        {"AboutDialog.Logo", ":icons/skin/qbittorrent-tray.svg"},
+        {"AddNewTorrentDialog.TorrentContentView.RenameAction", ":icons/qbt-theme/edit-rename.svg"},
+        {"CategoryFilter.ContextMenu.AddAction", ":icons/qbt-theme/list-add.svg"},
+        {"CategoryFilter.ContextMenu.AddSubcategoryAction", ":icons/qbt-theme/list-add.svg"},
```
Will remove "ContextMenu".

jagannatharjun replied:
Also should I change "CategoryFilter.Unused.RemoveAction" to "CategoryFilter.UnusedCatergories.RemoveAction"

glassez replied:
`UnusedCatergories` is better, IMO.

jagannatharjun commented on src/gui/uithememanager.cpp:
```diff
@@ -30,17 +30,489 @@
+    {
+        // load icon config from file
+        QFile iconJsonMap(configFile);
+        if (!iconJsonMap.open(QIODevice::ReadOnly))
```
actually uncrustify removes it automatically.

glassez replied:
Then uncrustify misconfigured. You should follow qBittorrent Coding Guidelines.

jagannatharjun replied:
I have a habit of regularly formatting the file even while writing so it will be cumbersome to do it manually since on next format they're gone. So Can I convert these in single liners instead?

glassez replied:
>So Can I convert these in single liners instead?

If the lines don't get too long.
Otherwise you can try to customize uncrustify config to follow our coding style more fully.

jagannatharjun commented on src/gui/uithememanager.cpp:
```diff
@@ -30,17 +30,489 @@
+
+        function throws ThemeError if encountered with any error
+     */
+    QHash<QString, QString> loadIconConfig(const QString &configFile)
```
It should be a static member since It doesn't require anything from `uithememanager`.

glassez replied:
Sorry. I mean different things.
If we really don't need different icon config files then you can change it to non static member function that don't return its result but just modify corresponding class fields. Otherwise leave it as is (resolving other comments).

jagannatharjun replied:
should I do it inplace since there is only one iconconfig?

glassez replied:
Seems it degrades the readability of the calling code.
Splitting into sub-functions is still useful in terms of code structuring. If some code portion does "stand-alone" job, it makes sense to put it in a separate method (unless it's really trivial).

jagannatharjun replied:
> should I do it inplace since there is only one iconconfig?

I decided against it since this makes constructor too big.

jagannatharjun commented on src/gui/uithememanager.cpp:
```diff
@@ -95,50 +601,74 @@ QIcon UIThemeManager::getIcon(const QString &iconId) const
 }
 
-QIcon UIThemeManager::getIcon(const QString &iconId, const QString &fallback) const
+QIcon UIThemeManager::getIcon(const QString &iconId, const QString &fallbackSysThemeIcon) const
```
What shall be the name?
`MainMenu.Fallback.ConfigureAction` ,
`Fallback.MainMenu.ConfigureAction`,
`MainMenu.ConfigureAction.Fallback`,
`MainMenu.ConfigureAction2`

glassez replied:
>What shall be the name?

Sorry, don't understand you.
There is only one name for each IconID. But there should be the way to assign several mappings to it. You can use either "multi map" for system icon config or map each id to list of values.

jagannatharjun replied:
well doesn't this make pattern-based searching difficult?
`resolveId` will return more than one icon file?

jagannatharjun replied:
I am thinking to use "multi map".
it should only be limited to the system icon config?

glassez replied:
Of course. In case of custom icon theme author should provide truly existed files so no fallback is needed.

jagannatharjun replied:
since system icon config is internal mapping does it make sense to not do pattern based searching here.

jagannatharjun replied:
I am not doing pattern-based searching for system icon map.

glassez replied:
Yes, at a first glance.
But I need to look at the whole algorithm to answer exactly.

jagannatharjun commented on src/gui/rss/feedlistwidget.cpp:
```diff
@@ -235,16 +235,16 @@ QTreeWidgetItem *FeedListWidget::createItem(RSS::Item *rssItem, QTreeWidgetItem
     }
     else {
-        icon = UIThemeManager::instance()->getIcon("inode-directory");
+        icon = UIThemeManager::instance()->getIcon("RSS");
```
Where it is actually used? I mean I can't able to understand in which context this icon is actually used thats why such a weird name.

glassez commented:
Superseded by #12861.
